File: python/catzilla/startup_banner.py
"""
Catzilla Startup Banner System

This module provides a beautiful, professional startup banner that displays
server information in both production and development modes.
"""

import os
import platform
import sys
import time
from typing import Any, Dict, Optional


class StartupBanner:
    """Manages the startup banner display with dynamic content and styling"""

    def __init__(self, production: bool = False):
        self.production = production
        self.version = self._get_version()
        self.supports_unicode = self._check_unicode_support()
        self.supports_color = self._check_color_support()

    def print_banner(self, host: str, port: int, app_stats: Dict[str, Any]) -> None:
        """Print the startup banner with server information

        Args:
            host: Server bind host
            port: Server bind port
            app_stats: Application statistics dictionary
        """
        if self.production:
            self._print_production_banner(host, port, app_stats)
        else:
            self._print_development_banner(host, port, app_stats)

    def _print_production_banner(
        self, host: str, port: int, stats: Dict[str, Any]
    ) -> None:
        """Print clean production banner"""
        width = self._calculate_width(host, port, stats, production=True)

        # Use C banner if available for maximum performance
        try:
            from catzilla._catzilla import print_startup_banner

            # TODO: Call C function when available
            # print_startup_banner(stats, True)
            # return
        except (ImportError, AttributeError):
            pass

        # Fallback to Python implementation
        self._render_banner(host, port, stats, width, production=True)

    def _print_development_banner(
        self, host: str, port: int, stats: Dict[str, Any]
    ) -> None:
        """Print rich development banner with extra information"""
        width = self._calculate_width(host, port, stats, production=False)

        # Always use Python for development (more flexibility)
        self._render_banner(host, port, stats, width, production=False)

    def _render_banner(
        self, host: str, port: int, stats: Dict[str, Any], width: int, production: bool
    ) -> None:
        """Render the banner using Python"""
        # Box drawing characters
        if self.supports_unicode:
            top_left, top_right = "╔", "╗"
            bottom_left, bottom_right = "╚", "╝"
            horizontal = "═"
            vertical = "║"
        else:
            top_left, top_right = "+", "+"
            bottom_left, bottom_right = "+", "+"
            horizontal = "="
            vertical = "|"

        # Print top border
        print(top_left + horizontal * (width - 2) + top_right)

        # Print title
        mode_str = "PRODUCTION" if production else "DEVELOPMENT"
        title = f"🐱 Catzilla v{self.version} - {mode_str}"
        self._print_content_line(title, width, vertical)

        # Print URL
        url = f"http://{host}:{port}"
        self._print_content_line(url, width, vertical)

        # Print bind info
        bind_info = f"(bound on host {host} and port {port})"
        self._print_content_line(bind_info, width, vertical)

        # Empty line
        self._print_content_line("", width, vertical)

        # Print statistics
        self._print_kv_line("Routes", str(stats.get("route_count", 0)), width, vertical)
        self._print_kv_line(
            "Workers", str(stats.get("worker_count", 1)), width, vertical
        )

        jemalloc_status = (
            "Enabled" if stats.get("jemalloc_enabled", False) else "Disabled"
        )
        self._print_kv_line("jemalloc", jemalloc_status, width, vertical)

        # Development-only fields
        if not production:
            di_count = stats.get("di_service_count", 0)
            if di_count > 0:
                di_status = f"Enabled ({di_count} services)"
                self._print_kv_line("DI Container", di_status, width, vertical)

            validation_status = (
                "Enabled" if stats.get("auto_validation", False) else "Disabled"
            )
            self._print_kv_line("Auto Validation", validation_status, width, vertical)

            if stats.get("profiling_enabled", False):
                profiling_status = (
                    f"Enabled (interval: {stats.get('profiling_interval', 60)}s)"
                )
                self._print_kv_line("Profiling", profiling_status, width, vertical)

            if stats.get("background_tasks", False):
                self._print_kv_line("Background Tasks", "Enabled", width, vertical)

            self._print_kv_line("Debug Mode", "ON", width, vertical)

        self._print_kv_line("PID", str(stats.get("pid", os.getpid())), width, vertical)

        # Print bottom border
        print(bottom_left + horizontal * (width - 2) + bottom_right)

        # Add spacing
        print()

    def _print_content_line(self, content: str, width: int, border_char: str) -> None:
        """Print a content line with proper padding"""
        content_width = len(content)
        available_space = width - 4  # 2 borders + 2 padding

        if content_width > available_space:
            content = content[:available_space]
            content_width = available_space

        left_padding = 1
        right_padding = available_space - content_width - left_padding

        line = f"{border_char} {' ' * left_padding}{content}{' ' * right_padding} {border_char}"
        print(line)

    def _print_kv_line(
        self, key: str, value: str, width: int, border_char: str
    ) -> None:
        """Print a key-value line with dots"""
        available_space = width - 4  # 2 borders + 2 padding
        key_value_space = len(key) + len(value) + 2  # 2 for spaces
        dots_count = available_space - key_value_space

        if dots_count < 3:
            dots_count = 3

        dots = "." * dots_count
        line = f"{border_char} {key} {dots} {value} {border_char}"
        print(line)

    def _calculate_width(
        self, host: str, port: int, stats: Dict[str, Any], production: bool
    ) -> int:
        """Calculate optimal banner width based on content"""
        min_width = 60

        # Check various content lengths
        mode_str = "PRODUCTION" if production else "DEVELOPMENT"
        title = f"🐱 Catzilla v{self.version} - {mode_str}"
        url = f"http://{host}:{port}"
        bind_info = f"(bound on host {host} and port {port})"

        # Find maximum width needed
        max_width = max(min_width, len(title) + 4, len(url) + 4, len(bind_info) + 4)

        # Cap at reasonable maximum
        return min(max_width, 100)

    def _get_version(self) -> str:
        """Get Catzilla version string"""
        try:
            # Try to get version from package
            import catzilla

            if hasattr(catzilla, "__version__"):
                return catzilla.__version__
        except ImportError:
            pass

        # Fallback version
        return "0.1.0"

    def _check_unicode_support(self) -> bool:
        """Check if terminal supports Unicode box drawing"""
        # Check if we're writing to a terminal
        if not sys.stdout.isatty():
            return False

        # Check environment variables
        term = os.environ.get("TERM", "")
        lang = os.environ.get("LANG", "")
        lc_all = os.environ.get("LC_ALL", "")

        # Windows specific checks
        if platform.system() == "Windows":
            # Modern Windows Terminal supports Unicode
            if os.environ.get("WT_SESSION"):
                return True
            # Check for UTF-8 support
            try:
                import locale

                return "utf-8" in locale.getpreferredencoding().lower()
            except:
                return False

        # Unix-like systems
        if any(t in term for t in ["xterm", "screen", "tmux", "rxvt"]):
            return True

        if "utf-8" in lang.lower() or "utf-8" in lc_all.lower():
            return True

        # Conservative fallback
        return False

    def _check_color_support(self) -> bool:
        """Check if terminal supports ANSI colors"""
        if not sys.stdout.isatty():
            return False

        # Check common environment variables
        if os.environ.get("NO_COLOR"):
            return False

        if os.environ.get("FORCE_COLOR"):
            return True

        term = os.environ.get("TERM", "")
        if "color" in term or term in ["xterm", "xterm-256color", "screen", "tmux"]:
            return True

        # Windows specific
        if platform.system() == "Windows":
            return os.environ.get("WT_SESSION") is not None

        return True


def collect_app_stats(app) -> Dict[str, Any]:
    """Collect comprehensive application statistics for the banner

    Args:
        app: Catzilla application instance

    Returns:
        Dictionary containing application statistics
    """
    stats = {
        "route_count": 0,
        "worker_count": 1,
        "jemalloc_enabled": False,
        "auto_validation": False,
        "di_service_count": 0,
        "profiling_enabled": False,
        "profiling_interval": 60,
        "background_tasks": False,
        "pid": os.getpid(),
    }

    try:
        # Get route count
        if hasattr(app, "routes"):
            stats["route_count"] = len(app.routes())
        elif hasattr(app, "router") and hasattr(app.router, "routes"):
            stats["route_count"] = len(app.router.routes())

        # Get jemalloc status
        if hasattr(app, "has_jemalloc"):
            stats["jemalloc_enabled"] = app.has_jemalloc

        # Get auto-validation status
        if hasattr(app, "auto_validation"):
            stats["auto_validation"] = app.auto_validation

        # Get DI service count
        if hasattr(app, "di_container") and app.di_container:
            if hasattr(app.di_container, "list_services"):
                stats["di_service_count"] = len(app.di_container.list_services())

        # Get profiling status
        if hasattr(app, "memory_profiling"):
            stats["profiling_enabled"] = app.memory_profiling
        if hasattr(app, "memory_stats_interval"):
            stats["profiling_interval"] = app.memory_stats_interval

        # Get background tasks status
        if hasattr(app, "_task_system_enabled"):
            stats["background_tasks"] = app._task_system_enabled

    except Exception:
        # Ignore errors in stats collection
        pass

    return stats

File: src/core/startup_banner.c
#include "startup_banner.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#ifdef _WIN32
#include <windows.h>
#include <io.h>
#define isatty _isatty
#define fileno _fileno
#else
#include <unistd.h>
#endif

// Unicode box drawing characters
#define BOX_TOP_LEFT     "╔"
#define BOX_TOP_RIGHT    "╗"
#define BOX_BOTTOM_LEFT  "╚"
#define BOX_BOTTOM_RIGHT "╝"
#define BOX_HORIZONTAL   "═"
#define BOX_VERTICAL     "║"

// ASCII fallback characters
#define ASCII_TOP_LEFT     "+"
#define ASCII_TOP_RIGHT    "+"
#define ASCII_BOTTOM_LEFT  "+"
#define ASCII_BOTTOM_RIGHT "+"
#define ASCII_HORIZONTAL   "="
#define ASCII_VERTICAL     "|"

void catzilla_print_startup_banner(const catzilla_server_stats_t* stats, bool production_mode) {
    if (!stats) return;

    bool use_unicode = catzilla_terminal_supports_unicode();
    int width = catzilla_calculate_banner_width(stats, production_mode);

    char line_buffer[512];
    char content_buffer[256];

    // Print top border
    catzilla_render_box_line(line_buffer, sizeof(line_buffer), width, true, use_unicode);
    printf("%s\n", line_buffer);

    // Print title line
    const char* mode_str = production_mode ? "PRODUCTION" : "DEVELOPMENT";
    snprintf(content_buffer, sizeof(content_buffer), "🐱 Catzilla v%s - %s", stats->version, mode_str);
    catzilla_render_content_line(line_buffer, sizeof(line_buffer), width, content_buffer, use_unicode);
    printf("%s\n", line_buffer);

    // Print URL line
    snprintf(content_buffer, sizeof(content_buffer), "http://%s:%d", stats->bind_host, stats->bind_port);
    catzilla_render_content_line(line_buffer, sizeof(line_buffer), width, content_buffer, use_unicode);
    printf("%s\n", line_buffer);

    // Print bind info line
    snprintf(content_buffer, sizeof(content_buffer), "(bound on host %s and port %d)", stats->bind_host, stats->bind_port);
    catzilla_render_content_line(line_buffer, sizeof(line_buffer), width, content_buffer, use_unicode);
    printf("%s\n", line_buffer);

    // Print empty line
    catzilla_render_content_line(line_buffer, sizeof(line_buffer), width, "", use_unicode);
    printf("%s\n", line_buffer);

    // Print statistics
    snprintf(content_buffer, sizeof(content_buffer), "%d", stats->route_count);
    catzilla_render_kv_line(line_buffer, sizeof(line_buffer), width, "Routes", content_buffer, use_unicode);
    printf("%s\n", line_buffer);

    snprintf(content_buffer, sizeof(content_buffer), "%d", stats->worker_count);
    catzilla_render_kv_line(line_buffer, sizeof(line_buffer), width, "Workers", content_buffer, use_unicode);
    printf("%s\n", line_buffer);

    const char* jemalloc_status = stats->jemalloc_enabled ? "Enabled" : "Disabled";
    catzilla_render_kv_line(line_buffer, sizeof(line_buffer), width, "jemalloc", jemalloc_status, use_unicode);
    printf("%s\n", line_buffer);

    // Development-only fields
    if (!production_mode) {
        if (stats->di_service_count > 0) {
            snprintf(content_buffer, sizeof(content_buffer), "Enabled (%d services)", stats->di_service_count);
            catzilla_render_kv_line(line_buffer, sizeof(line_buffer), width, "DI Container", content_buffer, use_unicode);
            printf("%s\n", line_buffer);
        }

        const char* validation_status = stats->auto_validation ? "Enabled" : "Disabled";
        catzilla_render_kv_line(line_buffer, sizeof(line_buffer), width, "Auto Validation", validation_status, use_unicode);
        printf("%s\n", line_buffer);

        if (stats->profiling_enabled) {
            snprintf(content_buffer, sizeof(content_buffer), "Enabled (interval: %ds)", stats->profiling_interval);
            catzilla_render_kv_line(line_buffer, sizeof(line_buffer), width, "Profiling", content_buffer, use_unicode);
            printf("%s\n", line_buffer);
        }

        catzilla_render_kv_line(line_buffer, sizeof(line_buffer), width, "Debug Mode", "ON", use_unicode);
        printf("%s\n", line_buffer);
    }

    snprintf(content_buffer, sizeof(content_buffer), "%d", (int)stats->pid);
    catzilla_render_kv_line(line_buffer, sizeof(line_buffer), width, "PID", content_buffer, use_unicode);
    printf("%s\n", line_buffer);

    // Print bottom border
    catzilla_render_box_line(line_buffer, sizeof(line_buffer), width, false, use_unicode);
    printf("%s\n", line_buffer);

    // Print newline for spacing
    printf("\n");
}

void catzilla_render_box_line(char* buffer, size_t buffer_size, int width, bool top_border, bool use_unicode) {
    if (!buffer || buffer_size == 0 || width < 3) return;

    const char* left_char = use_unicode ?
        (top_border ? BOX_TOP_LEFT : BOX_BOTTOM_LEFT) :
        (top_border ? ASCII_TOP_LEFT : ASCII_BOTTOM_LEFT);
    const char* right_char = use_unicode ?
        (top_border ? BOX_TOP_RIGHT : BOX_BOTTOM_RIGHT) :
        (top_border ? ASCII_TOP_RIGHT : ASCII_BOTTOM_RIGHT);
    const char* fill_char = use_unicode ? BOX_HORIZONTAL : ASCII_HORIZONTAL;

    int pos = 0;

    // Left border
    pos += snprintf(buffer + pos, buffer_size - pos, "%s", left_char);

    // Fill middle
    for (int i = 0; i < width - 2 && pos < (int)buffer_size - 1; i++) {
        pos += snprintf(buffer + pos, buffer_size - pos, "%s", fill_char);
    }

    // Right border
    if (pos < (int)buffer_size - 1) {
        snprintf(buffer + pos, buffer_size - pos, "%s", right_char);
    }
}

void catzilla_render_content_line(char* buffer, size_t buffer_size, int width, const char* content, bool use_unicode) {
    if (!buffer || buffer_size == 0 || width < 3) return;

    const char* border_char = use_unicode ? BOX_VERTICAL : ASCII_VERTICAL;

    // Calculate padding
    int content_len = content ? (int)strlen(content) : 0;
    int available_space = width - 4; // 2 for borders, 2 for padding
    int left_padding = 1;
    int right_padding = available_space - content_len - left_padding;

    if (right_padding < 0) {
        right_padding = 0;
        content_len = available_space - left_padding;
    }

    snprintf(buffer, buffer_size, "%s%*s%.*s%*s%s",
             border_char,
             left_padding, "",
             content_len, content ? content : "",
             right_padding, "",
             border_char);
}

void catzilla_render_kv_line(char* buffer, size_t buffer_size, int width, const char* key, const char* value, bool use_unicode) {
    if (!buffer || buffer_size == 0 || width < 3 || !key || !value) return;

    const char* border_char = use_unicode ? BOX_VERTICAL : ASCII_VERTICAL;

    // Calculate spacing
    int key_len = (int)strlen(key);
    int value_len = (int)strlen(value);
    int available_space = width - 4; // 2 for borders, 2 for padding
    int dots_space = available_space - key_len - value_len - 2; // 2 for spaces around dots

    if (dots_space < 3) dots_space = 3; // Minimum dots

    char dots[64];
    memset(dots, '.', sizeof(dots) - 1);
    dots[dots_space < 63 ? dots_space : 63] = '\0';

    snprintf(buffer, buffer_size, "%s %s %s %s %s",
             border_char, key, dots, value, border_char);
}

bool catzilla_terminal_supports_unicode(void) {
    // Check if we're writing to a terminal
    if (!isatty(fileno(stdout))) {
        return false;
    }

#ifdef _WIN32
    // On Windows, check console mode and codepage
    HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
    if (hConsole == INVALID_HANDLE_VALUE) return false;

    CONSOLE_SCREEN_BUFFER_INFO csbi;
    if (!GetConsoleScreenBufferInfo(hConsole, &csbi)) return false;

    // Check if we're in a modern Windows terminal
    DWORD mode;
    if (GetConsoleMode(hConsole, &mode)) {
        // Check for virtual terminal sequences support
        if (mode & ENABLE_VIRTUAL_TERMINAL_PROCESSING) {
            return true;
        }
    }

    // Check codepage for UTF-8 support
    return GetConsoleOutputCP() == CP_UTF8;
#else
    // On Unix-like systems, check environment variables
    const char* term = getenv("TERM");
    const char* lang = getenv("LANG");
    const char* lc_all = getenv("LC_ALL");

    // Most modern terminals support Unicode
    if (term && (strstr(term, "xterm") || strstr(term, "screen") ||
                strstr(term, "tmux") || strstr(term, "rxvt"))) {
        return true;
    }

    // Check locale for UTF-8
    if ((lang && strstr(lang, "UTF-8")) || (lc_all && strstr(lc_all, "UTF-8"))) {
        return true;
    }

    // Conservative fallback
    return false;
#endif
}

int catzilla_calculate_banner_width(const catzilla_server_stats_t* stats, bool production_mode) {
    if (!stats) return 60;

    // Calculate width based on content
    int min_width = 60;

    // Check various content lengths
    char temp_buffer[256];

    // Title line
    snprintf(temp_buffer, sizeof(temp_buffer), "🐱 Catzilla v%s - %s",
             stats->version, production_mode ? "PRODUCTION" : "DEVELOPMENT");
    int title_width = (int)strlen(temp_buffer) + 4; // 2 borders + 2 padding

    // URL line
    snprintf(temp_buffer, sizeof(temp_buffer), "http://%s:%d", stats->bind_host, stats->bind_port);
    int url_width = (int)strlen(temp_buffer) + 4;

    // Bind info line
    snprintf(temp_buffer, sizeof(temp_buffer), "(bound on host %s and port %d)", stats->bind_host, stats->bind_port);
    int bind_width = (int)strlen(temp_buffer) + 4;

    // Find maximum width needed
    int max_width = min_width;
    if (title_width > max_width) max_width = title_width;
    if (url_width > max_width) max_width = url_width;
    if (bind_width > max_width) max_width = bind_width;

    // Cap at reasonable maximum
    if (max_width > 100) max_width = 100;

    return max_width;
}
